aws/monitorinstances.go

```diff
@@ -0,0 +1,308 @@
+	stats, err := svc.GetMetricStatistics(&cloudwatch.GetMetricStatisticsInput{
+		Namespace:  aws.String("AWS/EC2"),
+		MetricName: aws.String("CPUUtilization"),
+		StartTime:  aws.Time(start),
+		EndTime:    aws.Time(end),
+		Period:     aws.Int64(int64(86400 * 30)),
+		Statistics: []*string{aws.String("Average"), aws.String("Maximum")},
+		Dimensions: dimensions,
+	})
+	if err != nil {
+		logger.Error("Fetch CPU stats from CloudWatch failed", err)
+		return 0, 0
+	} else if len(stats.Datapoints) > 0 {
+		return aws.Float64Value(stats.Datapoints[0].Average), aws.Float64Value(stats.Datapoints[0].Maximum)
+	} else {
+		return 0, 0
+	}
+}
+
+// fetchInstancesList, firstly, calls DescribeInstances from AWS to list all the instances in a region.
+// Secondly, it generates an InstanceInfo filled by DescribeInstances, getAccountID and getInstanceStats.
+// Finally, it sends the InstanceInfo generated in a chan.
+func fetchInstancesList(creds *credentials.Credentials, region string, account string,
+	logger jsonlog.Logger, instanceInfoChan chan InstanceInfo) {
+	defer close(instanceInfoChan)
+	sess := session.Must(session.NewSession(&aws.Config{
+		Credentials: creds,
+		Region:      aws.String(region),
+	}))
+	svc := ec2.New(sess)
+	instances, err := svc.DescribeInstances(nil)
+	if err != nil {
+		logger.Error("Describe instances failed", err)
+		return
+	}
+	now := time.Now()
+	hour := time.Date(now.Year(), now.Month(), now.Day(), now.Hour()-1, 0, 0, 0, now.Location())
+	for _, reservation := range instances.Reservations {
+		for _, instance := range reservation.Instances {
+			CPUAverage, CPUPeak := getInstanceStats(aws.StringValue(instance.InstanceId), sess, region, logger)
+			instanceInfoChan <- InstanceInfo{
+				Account:    account,
+				Service:    "EC2",
+				ID:         aws.StringValue(instance.InstanceId),
+				Region:     aws.StringValue(instance.Placement.AvailabilityZone),
+				Hour:       hour,
+				KeyPair:    aws.StringValue(instance.KeyName),
+				Tags:       getInstanceTag(instance.Tags),
+				Type:       aws.StringValue(instance.InstanceType),
+				CPUAverage: CPUAverage,
+				CPUPeak:    CPUPeak,
+			}
+		}
+	}
+}
+
+// fetchRegionsList fetchs the regions list from AWS and returns an array of their name.
+func fetchRegionsList(creds *credentials.Credentials, logger jsonlog.Logger) []string {
+	sess := session.Must(session.NewSession(&aws.Config{
+		Credentials: creds,
+		Region:      aws.String(defaultRegion),
+	}))
+	svc := ec2.New(sess)
+	regions, err := svc.DescribeRegions(nil)
+	if err != nil {
+		logger.Error("Describe regions failed", err)
+		return []string{}
+	}
+	res := make([]string, 0, 10)
+	for _, region := range regions.Regions {
+		res = append(res, aws.StringValue(region.RegionName))
+	}
+	return res
+}
+
+// getPropertiesMapping uses the reflection of the struct InstanceInfo and
+// InstanceTag to get the mapping of the ElasticSearch documents.
+func getPropertiesMapping(field interface{}) map[string]interface{} {
```

Comment:
A function that takes `field interface{}` and returns `map[string]interface{}` is really not explicit.
Either stop using `interface{}` or explain them in the documentation